### PR TITLE
MQTT notifications

### PR DESCRIPTION
**Pull request by fgalan**
Issue #3001

Sub-PRs included in this PR:

- [x] PR #3194 - the work done by @McMutton
- [x] PR #3889 - some small fixes after merging the former (pending comments in PR #3194, code style, functional tests & accumulator script, etc.)
- [x] PR #3897 - CRUD of the new information in the sub document, taking into account new fields for the csubs cache entries)
- [x] PR #3899 - Effective use of new information (topic, QoS, etc.) but still in mono-MQTTbroker way
- [x] PR #3903 - Implement multi-MQTTBroker (removing the CLI fields `-mqttHost` and `-mqttPort`)
- [x] PR #3904 - Improve MQTT subscription parsing
- [x] PR #3905 - Upgrade to last current mosquitto version (2.0.11)
- [x] PR #3912 - Documentation

Check also before merging:

- [x] CNR - PR #3912
- [x] Documenation (user, admin, devel) - PR #3912
- [x] Check if RPM builds (or if modifications to .spec are needed) - see [this comment](https://github.com/telefonicaid/fiware-orion/pull/3905#issuecomment-895154304)
- [x] Solve all pending FIXME PR comments - checked at commit c3105c7, no "FIXME PR" occurrence

### CODE REVIEW

File: test/functionalTest/cases/3001_mqtt/subscribe_and_update.test [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+03. Update entity attribute A1 to trigger notification
+======================================================
+HTTP/1.1 204 No Content
+Content-Length: 0
+Fiware-Correlator: REGEX([0-9a-f\-]{36})
+Date: REGEX(.*)
+
+
+
```
Comment (kzangeli) [inline test/functionalTest/cases/3001_mqtt/subscribe_and_update.test:130]: Where is the dump of the MQTT notification?
Without that, the test is a little bit worthless ...

Comment (fgalan) [reply test/functionalTest/cases/3001_mqtt/subscribe_and_update.test:130]: Check sub-PR https://github.com/telefonicaid/fiware-orion/pull/3889

**Comment by fgalan** [comment]
Cheat sheet, to start the MQTT broker:

```
docker run -d -p 1883:1883   -l mosquitto eclipse-mosquitto
```

File: test/functionalTest/cases/3001_mqtt/subscribe_and_update.test (hunk 1 above)
Comment (kzangeli) [reply test/functionalTest/cases/3001_mqtt/subscribe_and_update.test:130]: PR #3889 ... that's a PR to a feature branch, while this is a PR to "master" ...

Comment (kzangeli) [reply test/functionalTest/cases/3001_mqtt/subscribe_and_update.test:130]: Did see this though:
```
echo "# 04. Dump accumulator to see notification"
echo "=========================================="
mqttAccumulatorDump
echo
echo
```
No change in the --REGEXPECT-- section ...

Comment (fgalan) [reply test/functionalTest/cases/3001_mqtt/subscribe_and_update.test:130]: Yes. What I mean is that your comment is solved (or in the process to be solved ;) in that sub-PR of the present one.

Comment (fgalan) [reply test/functionalTest/cases/3001_mqtt/subscribe_and_update.test:130]: > No change in the --REGEXPECT-- section ...

PR #3889 is not yet completed.

File: test/functionalTest/harnessFunctions.sh [hunk 2]
```diff
@@ -771,6 +862,18 @@ function accumulatorStart()
...
 
+# ------------------------------------------------------------------------------
+#
+# mqttAccumulatorDump
+#
+function mqttAccumulatorDump()
+{
+  # The dump message is not necessary but nice to have to emphasize intention
+  mosquitto_pub -h $bindIp -t dump_mqtt -m "dump"
+}
```
Comment (fgalan) [inline test/functionalTest/harnessFunctions.sh:873]: I'm realizing this is not going to work.

accumulatorDump() get the output in the call to the accumulator API, which is based in HTTP so it's synchronous (in other words, the response contains the dump). However, in the case of MQTT, just publishing on a given topic (dump_mqtt topic) is not enough: the testHarness.sh script should keep listening at MQTT in asynchronous way to get the response published under some other topic and this is hard to do.

A best approach is probably include the MQTT listening functionality now in accumulator-mqtt.py in accumulator-server.py, preserving the HTTP based API to get the dump. I mean, accumulator-server.py would be able to listen to both HTTP notifications and MQTT notification (some flags in the CLI will be introduced for that), but the way of getting the notification dump will be the same (HTTP based API, wrapped by accumulatorDump() in harnessFunctions.sh)

File: test/functionalTest/cases/3001_mqtt/subscribe_and_update.test (hunk 1 above)
Comment (fgalan) [reply test/functionalTest/cases/3001_mqtt/subscribe_and_update.test:130]: Fixed by PR https://github.com/telefonicaid/fiware-orion/pull/3889

File: test/functionalTest/harnessFunctions.sh (hunk 2 above)
Comment (fgalan) [reply test/functionalTest/harnessFunctions.sh:873]: Fixed by PR https://github.com/telefonicaid/fiware-orion/pull/3889

**Comment by fgalan** [comment]
@fisuda this PR adds new .md documents and modify others. In addition, I understand some modifications should be done at doc/manuals/admin/build_source.md#ubuntu-2004-lts section with regards the installation of the new mosquitto library (and maybe some other similar changes in pieces of documentation you mantain).

It would be great if you could do a PR for that. Do it as PR for the feature/mqtt_notifications_poc branch, please.

Thanks!

**Comment by fisuda** [comment]
Noted with thanks. 
I have to set up an environment to test this PR on Ubuntu. First of all, I will update Japanese documentation except build_source.md#ubuntu-2004-lts section.

**Comment by fgalan** [comment]
Fails in commit bae2a66 are not meaningful. They are due to the current CI image has been built at PR https://github.com/telefonicaid/fiware-orion/pull/3880 and that one doesn't have yet the mosquitto library installed.

**Review by mapedraza — APPROVED** [review]
LGTM

**Pull request merged**